File: Dynamic Programming/Word Break/word_break.cpp
// DP. dp[i] = true if s[:i] can be segmented to words. dp[i] = true if dp[j] && s[j:i] in wordDict for any j < i. return dp[n].
class WordBreak {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        std::vector<bool> dp(s.length() + 1, false);
        std::unordered_set<std::string> words(wordDict.begin(), wordDict.end());
        dp[0] = true;

        for (int i = 1; i <= s.length(); ++i) {
            for (int j = 0; j < i; ++j) {
                if (dp[j] && words.contains(s.substr(j, i - j))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[s.length()];
    }
};

File: Dynamic Programming/Word Break/WordBreak.java
// DP. dp[i] = true if s[:i] can be segmented into words which occurs if dp[j] is true for j < i and s[j:i] is a valid word. return dp[n].
public class WordBreak {
	public boolean wordBreak(String s, List<String> wordDict) {
		int n = s.length();
		Set<String> wordSet = new HashSet<>(wordDict);
		boolean[] dp = new boolean[n + 1];
		dp[0] = true;
		for (int i = 1; i <= n; i++) {
			for (int j = 0; j < i; j++) {
				if (dp[j] && wordSet.contains(s.substring(j, i))) {
					dp[i] = true;
					break;
				}
			}
		}
		return dp[n];
	}
}
